src/sentry/notifications/helpers.py

```diff
@@ -0,0 +1,150 @@
+from collections import defaultdict
+from typing import Iterable, Mapping, Optional
+
+from sentry.notifications.legacy_mappings import get_legacy_value
+from sentry.notifications.types import (
+    NotificationScopeType,
+    NotificationSettingOptionValues,
+    NotificationSettingTypes,
+)
+
+
+def should_user_be_notified(
+    notification_settings_by_user: Mapping[
+        any, Mapping[NotificationScopeType, NotificationSettingOptionValues]
+    ],
+    user,
+) -> bool:
+    """
+    Given a mapping of default and specific notification settings by user,
+    determine if a user should receive an ISSUE_ALERT notification.
+    """
+    specific_scope = get_scope_type(NotificationSettingTypes.ISSUE_ALERTS)
+    notification_settings_option = notification_settings_by_user.get(user, {})
+    notification_setting_option = notification_settings_option.get(
+        specific_scope
+    ) or notification_settings_option.get(NotificationScopeType.USER)
+    value = getattr(
+        notification_setting_option,
+        "value",
+        NotificationSettingOptionValues.ALWAYS,
+    )
+
+    return value == NotificationSettingOptionValues.ALWAYS
+
+
+def should_be_participating(
+    user,
+    subscriptions_by_user_id: Mapping[int, any],
+    notification_settings_by_user: Mapping[
+        any, Mapping[NotificationScopeType, NotificationSettingOptionValues]
+    ],
+) -> bool:
+    """
+    Given a mapping of users to subscriptions and a mapping of default and
+    specific notification settings by user, determine if a user should receive
+    a WORKFLOW notification.
+    """
+    specific_scope = get_scope_type(NotificationSettingTypes.WORKFLOW)
+    notification_setting_option = (
+        notification_settings_by_user.get(user)[specific_scope]
+        or notification_settings_by_user.get(user)[NotificationScopeType.USER]
+    )
+    value = getattr(
+        notification_setting_option,
+        "value",
+        NotificationSettingOptionValues.SUBSCRIBE_ONLY,
+    )
+    if value == NotificationSettingOptionValues.NEVER.value:
+        return False
+
+    if value == NotificationSettingOptionValues.ALWAYS.value:
+        return True
+
+    subscription = subscriptions_by_user_id.get(user.id)
+    return subscription and subscription.is_active
+
+
+def transform_to_notification_settings_by_user(
+    notification_settings: Iterable,
+    users: Iterable,
+) -> Mapping[any, Mapping[NotificationScopeType, NotificationSettingOptionValues]]:
+    """
+    Given a unorganized list of notification settings, create a mapping of
+    users to a map of notification scopes to setting values.
+    """
+    actor_mapping = {user.actor: user for user in users}
+    notification_settings_by_user = defaultdict(dict)
+    for notification_setting in notification_settings:
+        user = actor_mapping.get(notification_setting.target)
```

Comment:
Is it valid/possible for user to be `None` here?

Reply:
We looked up the NotificationSettings by this `actor_id` and because we got the ID from the list of users, it should always be in that list. 
I made it paranoid to handle a missing User to appease the type checker.